src/shared/Microsoft.Git.CredentialManager.UI/Dispatcher.cs

```diff
@@ -0,0 +1,223 @@
+        /// </summary>
+        /// <param name="work">Work to be run.</param>
+        public void Post(Action work)
+        {
+            Task _ = InvokeAsync(work);
+        }
+
+        /// <summary>
+        /// Execute work to be run on the thread associated with this dispatcher and wait
+        /// synchronously until the work is complete.
+        /// </summary>
+        /// <param name="work">Work to be run.</param>
+        public Task InvokeAsync(Action work)
+        {
+            var tcs = new TaskCompletionSource<object>();
+            _queue.AddJob(new DispatcherJob(work, tcs));
+            return tcs.Task;
+        }
+
+        public Task<TResult> InvokeAsync<TResult>(Func<TResult> work)
+        {
+            var tcs = new TaskCompletionSource<TResult>();
+            _queue.AddJob(new DispatcherJob<TResult>(work, tcs));
+            return tcs.Task;
+        }
+
+        private interface IDispatcherJob
+        {
+            void Execute();
+        }
+
+        private class DispatcherJob : IDispatcherJob
+        {
+            private readonly Action _work;
+            private readonly TaskCompletionSource<object> _tcs;
+
+            public DispatcherJob(Action work, TaskCompletionSource<object> tcs)
+            {
+                _work = work;
+                _tcs = tcs;
+            }
+
+            public void Execute()
+            {
+                _work();
+                _tcs?.SetResult(null);
+            }
+        }
+
+        private class DispatcherJob<TResult> : IDispatcherJob
+        {
+            private readonly Func<TResult> _work;
+            private readonly TaskCompletionSource<TResult> _tcs;
+
+            public DispatcherJob(Func<TResult> work, TaskCompletionSource<TResult> tcs)
+            {
+                _work = work;
+                _tcs = tcs;
+            }
+
+            public void Execute()
+            {
+                TResult result = _work();
+                _tcs?.SetResult(result);
+            }
+        }
+
+        private class DispatcherJobQueue
+        {
+            private readonly Queue<IDispatcherJob> _queue = new();
+
+            private enum State
+            {
+                NotStarted,
+                Started,
+                Stopping,
+                Stopped,
+            }
+
+            private State _state = State.NotStarted;
+
+            public void Run()
+            {
+                lock (_queue)
+                {
+                    switch (_state)
+                    {
+                        case State.Started:
+                            throw new InvalidOperationException("Dispatcher has already started.");
+                        case State.Stopping:
+                            throw new InvalidOperationException("Dispatcher is shutting down.");
+                        case State.Stopped:
+                            throw new InvalidOperationException("Dispatcher has shut down.");
+                    }
```

Comment:
Could these `throws` be updated to helper methods? I just see that you have duplicate strings, even though the `switch` statements are slightly different.

Even better would be to pull the full `switch` into a helper that takes a list of possible "acceptable" states, but otherwise throws the appropriate error. If `state` becomes a flag-type enum, then you could pass it like this:

(here)
`CheckState(validStates: State.NotStarted);`

(in `AddJob`)
`CheckState(validStates: State.NotStarted | State.Started);`

Reply:
I think I'll leave this as is for now. I think the current switches are clear, and simple enough that adding another method doesn't add much. The error strings are actually slightly different between the Start and Stop methods.